bottlecap/src/lifecycle/listener.rs

```diff
@@ -33,33 +34,64 @@ const AGENT_PORT: usize = 8124;
 
 pub struct Listener {
     pub invocation_processor: Arc<Mutex<InvocationProcessor>>,
+    pub shutdown_token: CancellationToken,
 }
 
+type ListenerState = (Arc<Mutex<InvocationProcessor>>, Arc<Mutex<JoinSet<()>>>);
+
 impl Listener {
+    pub fn new(invocation_processor: Arc<Mutex<InvocationProcessor>>) -> Self {
+        let shutdown_token = CancellationToken::new();
+        Self {
+            invocation_processor,
+            shutdown_token,
+        }
+    }
+
+    #[must_use]
+    pub fn get_shutdown_token(&self) -> CancellationToken {
+        self.shutdown_token.clone()
+    }
+
     pub async fn start(&self) -> Result<(), Box<dyn std::error::Error>> {
         let port = u16::try_from(AGENT_PORT).expect("AGENT_PORT is too large");
         let addr = SocketAddr::from(([127, 0, 0, 1], port));
         let listener = TcpListener::bind(&addr).await?;
 
-        let router = self.make_router();
+        let tasks = Arc::new(Mutex::new(JoinSet::new()));
```

Comment:
Oh we need a mutex for the joinset because we pass it between threads in the handler method tasks?

Reply:
Yeah correct, so multiple handlers can append tasks, but it should be pretty fast, that way the graceful shutdown can finish correctly in case of errors